### PR TITLE
Refactor tracing implements

### PR DESCRIPTION
**Pull request by XSAM**
Split from #275

- Add tracing interceptor.
- Add opentracing provider.
- Add opentelemetry provider.

### CODE REVIEW

**Comment by XSAM** [comment]
@bwplotka I need some review to confirm the tracing interface since it implements parts of `interceptors.Reporter`. 😁

File: interceptors/tracing2/interceptors.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package tracing2
```
Comment (bwplotka) [inline interceptors/tracing2/interceptors.go:1]: Why not modifying `tracing` directly?

Comment (XSAM) [reply interceptors/tracing2/interceptors.go:1]: It makes the code easier to review.

I will rename it back to `tracing` in the last commit.

**Comment by XSAM** [comment]
Well, `opentracing` and `OTEL (opentelemtry)` do have a familiar interface of span to record events. 

`opentracing`: LogFields, `OTEL`: AddEvent.

But the problem is the behavior from two libs are not the same while handling `PostMsgSend` and `PostMsgReceive`.

For instance, when invoking `PostMsgSend`, `opentracing` will do nothing with `LogFields` since it doesn't require to record gRPC message info, but `OTEL` will record that info via `AddEvent`.

Furthermore, attributes which been using in the event are different.

`OTEL`:

```go
span.AddEvent(ctx, "message",
	kv.Key("message.type").String("SENT"),
	kv.Key("message.id").Int(id),
	kv.Key("message.uncompressed_size").Int(proto.Size(body)),
)
```

So, I think there are two solutions to this.

1. Add `AddEvent` function for span interface. But it will force `opentracing` to add the same key and attributes to events like `OTEL`, or `opentracing` have to do some weird mapping to drop these non-standard events.

2. Add `PostMsgSend` and `PostMsgReceive` for span interface. Weird but more flexible.

-----

- https://github.com/opentracing/specification/blob/master/semantic_conventions.md
- https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md#events

**Comment by bwplotka** [comment]
Hi @XSAM Let's get back to this!

I feel like option number 2 is too similar too reporter, so kind of no point in having separate span interface, no? (: 

That's why we have two options I think

1. as your 1, we could try that.

3.  Actually have no common interface among those, just have 2 different providers/opentelemetry` and `providers/opentracing` ^^

**Comment by XSAM** [comment]
You're right, option 2 is too similar with `interceptors.Reporter`.

Now, I use option 1 as a solution.

**Comment by yashrsharma44** [comment]
Seems like this PR would address #277 . @XSAM can you let us know if you are interested in working on rebasing this PR and meeting the acceptance criteria? Would be happy to take over this PR, otherwise 😄 

Related Issue - https://github.com/grpc-ecosystem/go-grpc-middleware/issues/277

**Comment by XSAM** [comment]
@yashrsharma44 I will continue to work on this PR. Thanks for the reminder. 😄

**Comment by XSAM** [comment]
I updated this PR. @yashrsharma44 @bwplotka PTAL ;)

**Comment by XSAM** [comment]
Any updates on this one?

**Comment by yashrsharma44** [comment]
Hi @XSAM!
Extremely sorry for the late review, I was busy with my internship requirements, so wasn't able to take a look at the PR.

* Can you split the changes for `opentracing` and `opentelemetry` in a separate PR?
* Also can you rebase your PR with the latest changes?

**Comment by bwplotka** [comment]
I think it's actually reviewable, in one place as those are new things mostly. But it will take time. 

Still, I think we should do it before v2?

**Comment by bwplotka** [comment]
Are you around @XSAM to work on it (rebase etc)? 

Also I am removing tracing here, to avoid modifications there that will have to be there as part of this Pr anyway (https://github.com/grpc-ecosystem/go-grpc-middleware/pull/394#issuecomment-894643059)

**Comment by XSAM** [comment]
> Are you around @XSAM to work on it (rebase etc)?
> 
> Also I am removing tracing here, to avoid modifications there that will have to be there as part of this Pr anyway ([#394 (comment)](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/394#issuecomment-894643059))

Sorry for the late response, I was busy recently.

I'm still working on it; will try to split these changes for clarity.

**Pull request closed without merging**